### ISSUE TITLE
Batch balancing repeatedly dispatches scalar tensor operations inside Python partitioning

### ISSUE THREAD

**Issue by ai-yang** [verl-project/verl#7279]
## System Info

- verl main commit: `2781c1e405fa80871a838bfdf9495b412df19c35`
- Intel Xeon Gold 6133, Linux x86_64
- Python 3.12.13
- PyTorch 2.13.0+cpu, one Torch thread, process pinned to CPU 2
- CUDA was unavailable, so the measurements below are CPU-only.

## Problem

PPO, V1 PPO, and SFT batch balancing calculate workload into a `torch.Tensor`, then pass its scalar tensors into the Python-native Karmarkar-Karp heap and Python sorting keys. Comparisons, heap operations, additions, and sort comparisons repeatedly dispatch PyTorch scalar operations.

The trainers use this controller-side path whenever batch balancing is enabled (the PPO and SFT trainer defaults enable it), so the overhead recurs during training rather than being a one-time setup cost.

## Reproduction

The benchmark includes workload calculation/conversion, Karmarkar-Karp partitioning, and the trainers' within-partition ordering. It uses seeded lengths from 64 through 8192, eight partitions, five warmups, randomized baseline/candidate order, and asserts identical final partition/order results.

Save the following harness as `/tmp/benchmark_balance.py`:

<details>
<summary>Benchmark source</summary>

```python
import gc
import random
import statistics
import time

import torch

from verl.utils.seqlen_balancing import calculate_workload, get_seqlen_balanced_partitions

torch.set_num_threads(1)
torch.set_num_interop_threads(1)
generator = torch.Generator().manual_seed(20260806)
order_rng = random.Random(20260806)


def balance_tensor(seqlens):
    workloads = calculate_workload(seqlens)
    partitions = get_seqlen_balanced_partitions(workloads, k_partitions=8, equal_size=True)
    for idx, partition in enumerate(partitions):
        partition.sort(key=lambda item: (workloads[item], item))
        partitions[idx] = partition[::2] + partition[1::2][::-1]
    return partitions


def balance_list(seqlens):
    workloads = calculate_workload(seqlens).tolist()
    partitions = get_seqlen_balanced_partitions(workloads, k_partitions=8, equal_size=True)
    for idx, partition in enumerate(partitions):
        partition.sort(key=lambda item: (workloads[item], item))
        partitions[idx] = partition[::2] + partition[1::2][::-1]
    return partitions


for batch_size, repetitions in ((256, 30), (1024, 20), (4096, 10)):
    seqlens = torch.randint(64, 8193, (batch_size,), dtype=torch.int64, generator=generator)
    expected = balance_tensor(seqlens)
    assert balance_list(seqlens) == expected
    for _ in range(5):
        balance_tensor(seqlens)
        balance_list(seqlens)

    timings = {"tensor": [], "list": []}
    jobs = [("tensor", balance_tensor), ("list", balance_list)]
    gc.disable()
    try:
        for _ in range(repetitions):
            order_rng.shuffle(jobs)
            for name, function in jobs:
                start = time.perf_counter_ns()
                result = function(seqlens)
                timings[name].append((time.perf_counter_ns() - start) / 1e6)
                assert result == expected
    finally:
        gc.enable()

    old = statistics.mean(timings["tensor"])
    new = statistics.mean(timings["list"])
    print(
        f"batch={batch_size} reps={repetitions} "
        f"tensor={old:.3f}+/-{statistics.stdev(timings['tensor']):.3f} ms "
        f"list={new:.3f}+/-{statistics.stdev(timings['list']):.3f} ms "
        f"speedup={old / new:.2f}x reduction={(old - new) / old * 100:.2f}%"
    )
```

</details>

Run it from the repository checkout:

```bash
taskset -c 2 .venv/bin/python /tmp/benchmark_balance.py
```

Measured on the system above:

| Batch | Repetitions | Tensor scalars | Python scalars | Speedup | Time reduction |
|---:|---:|---:|---:|---:|---:|
| 256 | 30 | 35.572 ± 0.736 ms | 1.598 ± 0.070 ms | 22.26× | 95.51% |
| 1,024 | 20 | 188.412 ± 1.437 ms | 7.569 ± 0.183 ms | 24.89× | 95.98% |
| 4,096 | 10 | 925.286 ± 5.659 ms | 33.940 ± 0.678 ms | 27.26× | 96.33% |

These are controller-side CPU timings, not end-to-end training throughput measurements. GPU-hosted training was not available for validation. Converting once should also avoid repeated device-to-host scalar synchronization for CUDA tensors, but that is an inference and is not claimed as a measured result.

## Expected behavior

Workloads should cross from Torch to native Python once before entering a Python-only heap/sort algorithm. The partition assignments and deterministic within-partition ordering must remain exactly unchanged.

**Comment by ai-yang** [comment]
### Proposed optimization, compatibility risks, and duplicate check

- **Implementation:** retain `calculate_workload()` for tensor consumers, add a one-time tensor-to-host-scalar boundary helper for Python-native partitioning, and use it in classic PPO, V1 PPO, and SFT. Skip the conversion entirely on the PrefixGrouper path where workload values are unused.
- **Expected scope:** the direct benefit is limited to controller-side batch balancing; it does not imply an end-to-end training speedup.
- **Correctness/compatibility risks addressed:** exact partition/order equivalence is asserted, integer and float32 inputs are covered, and NumPy 1.x/2.x float32 accumulation/tie-breaking behavior is preserved. There is no user-facing API or configuration change.
- **Duplicate distinction:** repeated open searches found only this issue and draft PR #7281. Issue #4152 concerns divisibility, PR #6025 grouped mini-batches, and other padding/sequence-balancing work does not remove scalar tensor dispatch at these three trainer boundaries.

The later 4×RTX 3090 ABBA result and final correctness checks are included directly in the [PR #7281 body](https://github.com/verl-project/verl/pull/7281).
